Add Jenkins assumable identities tutorial, and rudimentary inline %variable% editing (#717)
chainguard-dev/edu: content/chainguard/chainguard-enforce/iam-groups/identity-examples/enforce-jenkins-identity/index.md

The first section creates the identity itself.

```
resource "chainguard_identity" "jenkins" {
  parent_id   = chainguard_group.user-group.id
  name        = "jenkins"
  description = <<EOF
        This is an identity that authorizes Jenkins workflows
        for this repository to assume to interact with chainctl.
  EOF

    claim_match {
    audience = "%your-audience%"
    issuer   = "https://%your-domain%/oidc"
    subject  = "%your-subject%"
  }
}
```

First, this `%your-audience%` section creates a Chainguard Identity tied to the `chainguard_group` created by the `sample.tf` file; namely, the `example-group` group. The identity is named `jenkins` and has a brief description.

The most important part of this section is the `claim_match`. When the Jenkins workflow tries to assume this identity later on, it must present a token matching the `audience`, `issuer` and `subject` specified here in order to do so. The `audience` is the intended recipient of the issued token, while the `issuer` is the entity that creates the token. Finally, the `subject` is the entity (here, the Jenkins pipeline build) that the token represents.

> The `audience` and `issuer` fields use the setting from your configured Jenkins OIDC credential, in this case from the token configured on your Jenkins' `/manage/credentials/store/system/domain/_/` page. The following screenshot shows an example of these settings:

Author: Makes sense, anyone who knows how/where can skip this, otherwise, it will save them the work you had to go through to figure it out 😄

![Jenkins OICD token configuration page](jenkins-oidc-credential.png)

> For the subject, refer to your Jenkins repository OIDC settings page under the `/manage/configureSecurity/` URL. To find the subject field, scroll to the `OpenID Connect` section on the page, and then locate the `_sub_` field under the `Claim templates` pop-out settings. Refer to the following screenshot for an exmaple of these settings:

Reviewer: ```suggestion
For the subject, refer to your Jenkins repository OIDC settings page. You can find these by naivgating back to the **Manage Jenkins** landing page in your dashboard and clicking on **Security**. From there, scroll to the `OpenID Connect` section on the page, click on the **Claim templates** button, and locate the `sub` field. The `subject` value you should use will be the value in the **Value format** field under the first `sub` template. In the following example, the value to use is `jenkins-oidc-test`.
```

The next section will output the new identity's `id` value. This is a unique value that represents the identity itself.

```
output "jenkins-identity" {
  value = chainguard_identity.jenkins.id
}
```

The section after that looks up the `viewer` role.

```
data "chainguard_roles" "viewer" {
  name = "viewer"
}
```

The final section grants this role to the identity on the `example-group`.

```
resource "chainguard_rolebinding" "view-stuff" {
  identity = chainguard_identity.jenkins.id
  group    = chainguard_group.user-group.id
  role     = data.chainguard_roles.viewer.items[0].id
}
```

Run the following command to create this file with each of these sections. Be sure to change the `audience`, `subject`, and `subject` values to align with your own Jenkins pipeline.

```sh
cat > jenkins.tf <<EOF
resource "chainguard_identity" "jenkins" {
  parent_id   = chainguard_group.user-group.id
  name        = "jenkins"
  description = <<EOF
        This is an identity that authorizes Jenkins workflows
        for this repository to assume to interact with chainctl.
  EOF

    claim_match {
    audience        = "%your-audience%"
    issuer          = "https://%your-domain%/oidc"

>     subject_pattern = "%your-subject%"

Reviewer: ```suggestion
    subject = "%your-subject%"
```

  role     = data.chainguard_roles.viewer.items[0].id
}
EOF
```

Following that, your Terraform configuration will be ready. Now you can run a few `terraform` commands to create the resources defined in your `.tf` files.

## Creating Your Resources

First, run `terraform init` to initialize Terraform's working directory.

```sh
terraform init
```

Then run `terraform plan`. This will produce a speculative execution plan that outlines what steps Terraform will take to create the resources defined in the files you set up in the last section.

```sh
terraform plan
```

If the plan worked successfully and you're satisfied that it will produce the resources you expect, you can apply it.

```sh
terraform apply
```

Before going through with applying the Terraform configuration, this command will prompt you to confirm that you want it to do so. Enter `yes` to apply the configuration.

```
. . .

Plan: 4 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + jenkins-identity = (known after apply)

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value:
```

> After pressing `ENTER`, the command will complete and will output an `jenkins-identity` value. Note that you may receive a `PermissionDenied` error part way through the apply step. If so, run `chainctl auth login` once more, and then `terraform apply` again to resume creating the identity and resources.

Reviewer: ```suggestion
After pressing `ENTER`, the command will complete and will output an `jenkins-identity` value.
```

This sentence weirdly splits the output, if that makes sense

```
. . .

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.

Outputs:

jenkins-identity = "<your jenkins identity>"
```

This is the identity's [UIDP (unique identity path)](/chainguard/chainguard-enforce/reference/events/#uidp-identifiers), which you configured the `jenkins.tf` file to emit in the previous section. Note this value down, as you'll need it when you test this identity using a Jenkins workflow. If you need to retrieve this UIDP later on, though, you can always run the following `chainctl` command to obtain a list of the UIDPs of all your existing identities.

```sh
chainctl iam identities ls
```

> 

Reviewer: ```suggestion

Note that you may receive a `PermissionDenied` error part way through the apply step. If so, run `chainctl auth login` once more, and then `terraform apply` again to resume creating the identity and resources.

```

> You're now ready to create or edit a Jenkins pipeline job to test out this identity.

Reviewer: ```suggestion
You're now ready to create or edit a Jenkins pipeline to test out this identity.
```

## Testing the identity with a Jenkins pipeline

> To test the identity you created with Terraform in the previous section, create or edit a pipeline job. Ensure that the job is parameterized and uses the `OpenID Connect id token` credential type. Mark the parameter as required, and select your configured OIDC credential token as the `Default Value` for the parameter per the following screenshot:

Reviewer: Could you specify where they make this setting?

Now you can add the commands for testing the identity like `chainctl policy ls` in the following example:

```
sh '''
    wget -O chainctl "https://dl.enforce.dev/chainctl/latest/chainctl_linux_\$(uname -m)"
    chmod +x chainctl
    ./chainctl -v3 auth login --identity-token $token --identity <your jenkins identity>
    ./chainctl policy ls
'''
```

Save the job, and then build it using the `Build with Parameters` option.

Assuming everything works as expected, your pipeline will be able to assume the identity and run the `chainctl policy ls` command, returning the policy you created with the `sample.tf` file.

```
. . .

chainctl        	100%[===================>]  54.34M  6.78MB/s	in 13s

2023-05-17 13:19:45 (4.28 MB/s) - ‘chainctl’ saved [56983552/56983552]

Successfully exchanged token.
Valid! Id: 3f4ad8a9d5e63be71d631a359ba0a91dcade94ab/d3ed9c70b538a796
                         	ID                         	| 	NAME  	| DESCRIPTION |   MODE
------------------------------------------------------------+---------------+-------------+-----------

>   3f4ad8a9d5e63be71d631a359ba0a91dcade94ab/42b8649dc3a3ea66 | trust-any-cgr |         	| ENFORCED

Author: I think it will be, looks like a bunch of extra spaces made it into my output somehow. Thanks for the keen 👀

```

If you'd like to experiment further with this identity and what the pipeline can do with it, there are a few parts of this setup that you can tweak. For instance, if you'd like to give this identity different permissions you can change the role data source to the role you would like to grant.

```
data "chainguard_roles" "editor" {
  name = "editor"
}
```

You can also edit the pipeline itself to change its behavior. For example, instead of inspecting the policies the identity has access to, you could have the workflow inspect the groups.

```
	. . .

	- './chainctl iam groups ls'
```

Of course, the Jenkins pipeline will only be able to perform certain actions on certain resources, depending on what kind of access you grant it.

## Removing Sample Resources

To remove the resources Terraform created, you can run the `terraform destroy` command.

```sh
terraform destroy
```

This will destroy the sample policy, the role-binding, and the identity created in this guide. However, you'll need to destroy the `example-group` group yourself with `chainctl`.

> ```sh

Reviewer: so `terraform destroy` didn't destroy my policy or group, I had to delete them both manually. But I couldn't delete the group until the policy under was deleted first.

Author: That's an odd one, I'll give it another run through and check. If I can't get the policy to be deleted with terraform, at this point I'll just make a note to delete manually with chainctl.
